zcash_client_sqlite/src/wallet.rs

```diff
@@ -2758,13 +2758,21 @@ pub(crate) fn set_transaction_status(
     // is whether that transaction ends up being mined or expires.
     match status {
         TransactionStatus::TxidNotRecognized | TransactionStatus::NotInMainChain => {
-            // If the transaction is now expired, remove it from the retrieval queue.
+            // Remove the txid from the retrieval queue unless an unexpired transaction having that
+            // txid exists in the transactions table.
             if let Some(chain_tip) = chain_tip_height(conn)? {
                 conn.execute(
-                    "DELETE FROM tx_retrieval_queue WHERE txid IN (
-                        SELECT txid FROM transactions
-                        WHERE txid = :txid AND expiry_height < :chain_tip_minus_lookahead
-                    )",
+                    "DELETE FROM tx_retrieval_queue
+                     WHERE txid = :txid
+                     AND NOT EXISTS (
+                        SELECT id_tx
+                        FROM transactions t
+                        WHERE t.txid = :txid
+                        AND (
+                          t.expiry_height = 0
+                          OR t.expiry_height >= :chain_tip_minus_lookahead
```

Comment:
There's a state that is not adequately addressed by this version, which is when `expiry_height IS NULL`. It's unclear though what the right behavior should be in this case - we don't know the expiry height, so we don't have any reference for when we should accept the "not recognized / not in main chain" determination as final.